server/src-lib/Hasura/Backends/Postgres/Connection.hs

```diff
@@ -210,50 +228,123 @@ withTraceContext ctx = \case
           Q.unitQE defaultTxErrorHandler sql () False
      in LTTx $ setTraceContext >> tx
 
-instance Functor (LazyTx e) where
+instance (Functor m) => Functor (LazyTxT e m) where
   fmap f = \case
     LTErr e  -> LTErr e
     LTNoTx a -> LTNoTx $ f a
     LTTx tx  -> LTTx $ fmap f tx
 
-instance Applicative (LazyTx e) where
+instance (Monad m) => Applicative (LazyTxT e m) where
   pure = LTNoTx
 
-  LTErr e   <*> _        = LTErr e
-  LTNoTx f  <*> r        = fmap f r
-  LTTx _    <*> LTErr e  = LTErr e
-  LTTx txf  <*> LTNoTx a = LTTx $ txf <*> pure a
-  LTTx txf  <*> LTTx tx  = LTTx $ txf <*> tx
+  LTErr e   <*> _         = LTErr e
+  LTNoTx f  <*> r         = fmap f r
+  LTTx _    <*> LTErr e   = LTErr e
+  LTTx txf  <*> LTNoTx a  = LTTx $ txf <*> pure a
+  LTTx txf  <*> LTTx tx   = LTTx $ txf <*> tx
 
-instance Monad (LazyTx e) where
+instance (Monad m) => Monad (LazyTxT e m) where
   LTErr e >>= _  = LTErr e
   LTNoTx a >>= f = f a
   LTTx txa >>= f =
     LTTx $ txa >>= lazyTxToQTx . f
 
-instance MonadError e (LazyTx e) where
+instance (Monad m) => MonadError e (LazyTxT e m) where
   throwError = LTErr
   LTErr e  `catchError` f = f e
   LTNoTx a `catchError` _ = LTNoTx a
   LTTx txe `catchError` f =
     LTTx $ txe `catchError` (lazyTxToQTx . f)
 
-instance MonadTx (LazyTx QErr) where
-  liftTx = LTTx
+instance MonadTrans (LazyTxT e) where
+  lift = LTTx . lift
+
+instance (Tracing.MonadTrace m) => Tracing.MonadTrace (LazyTxT e m) where
+  trace t = \case
+    LTTx (Q.TxET tx) -> LTTx $ Q.TxET $ Tracing.trace t tx
+    v                -> v
+  currentContext  = lift Tracing.currentContext
+  currentReporter = lift Tracing.currentReporter
+  attachMetadata  = lift . Tracing.attachMetadata
 
-instance MonadTx (Q.TxE QErr) where
-  liftTx = id
+instance (MonadIO m) => MonadTx (LazyTxT QErr m) where
+  liftTx = LTTx . (hoist liftIO)
 
-instance MonadIO (LazyTx e) where
+instance (MonadIO m) => MonadTx (Q.TxET QErr m) where
+  liftTx = hoist liftIO
+
+instance (MonadIO m) => MonadIO (LazyTxT e m) where
   liftIO = LTTx . liftIO
 
-instance MonadBase IO (LazyTx e) where
+instance (MonadIO m) => MonadBase IO (LazyTxT e m) where
```

Comment:
Is this still required? This seems redundant if we already have `MonadIO` and `MonadBaseControl`.

Reply:
`MonadBase` is a precursor to `MonadBaseControl`. Its signature is as follows:
```haskell
class MonadBase b m => MonadBaseControl b m | m -> b where
```

Reply:
Oh, my bad!